connector/consumer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connector // import "go.opentelemetry.io/collector/connector"
+
+import (
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/fanoutconsumer"
+)
+
+// TracesConsumer feeds the first consumer.Traces in each of the specified pipelines.
+// Unlike other consumers, the connector consumer will expose its underlying consumers.
+type TracesConsumer interface {
```

Comment:
I think you can define only:

```golang
// MultiTracesConsumer is an optional interface implemented by consumer.Traces which fanout to multiple other consumers. (or FanoutTracesConsumer)
type MultiTracesConsumer interface {
  Consumers() map[component.ID]consumer.Traces
}
```

Since that is what go recommends for this cases.

Reply:
Separate question: 
Should this be here? Should we have it as part of "component" maybe?

Reply:
```
// TracesRouter is an optional interface implemented by consumer.Traces which fanout to multiple other consumers
type TracesRouter interface {
  Consumers() map[component.ID]consumer.Traces
}
```

> If we remove the consumer.Traces, then non-routing connectors will always have to handle their own fanout to all consumers. I think it's better if we can pass into the connector constructors the typical consumer.Traces interface, and require only non-routing connectors to be concerned with this optional method.

No because as the text said, that is an optional interface implemented by `consumer.Traces`. See https://github.com/open-telemetry/opentelemetry-collector/blob/main/connector/forwardconnector/forward.go#LL45-L53C2 you still receive `consumer.Traces`.

```golang
// createTracesToTraces creates a trace receiver based on provided config.
func createTracesToTraces(
	_ context.Context,
	set connector.CreateSettings,
	cfg component.Config,
	nextConsumer consumer.Traces,
) (connector.Traces, error) {
	router, ok = nextConsumer.(TracesRouter)
	if ok {
		// It means the next component is a router, so if I want to do manual routing I can access
		// router.Consumers()
	} else {
		next consumer is not a router means component is connected to only 1 pipeline.
	}
}
```

Reply:
> It means the next component is a router, so if I want to do manual routing I can access

We're trying to build a routing connector. The `TracesRouter` interface is meant to allow that, not indicate anything about the next component.

The comment about the interface being optional is not well stated, but my intention was to indicate that this interface can optionally be asserted, so that only routing connectors would assert this in order to gain the ability to route to specific consumers.

Reply:
But if user connected to only one pipeline?

Reply:
> But if user connected to only one pipeline?

If we are building a routing connector for example, my idea looks like this:
```go
func createTracesToTraces(
	_ context.Context,
	set connector.CreateSettings,
	cfg component.Config,
	nextConsumer consumer.Traces,
) (connector.Traces, error) {
	router, ok = nextConsumer.(TracesRouter)
	if !ok {
		// impossible because service ensures this is always `TracesRouter`, even though it passes in as `consumer.Traces`
	}
	if len(router.Consumers() < 2 {
		return fmt.Errorf("routing connector requires at least two downstream pipelines")
	// or maybe we decide it's ok to have only one. point is it can be checked either way
	}
}
```

Reply:
```
func createTracesToTraces(
	_ context.Context,
	set connector.CreateSettings,
	cfg component.Config,
	nextConsumer consumer.Traces,
) (connector.Traces, error) {
	router, ok = nextConsumer.(TracesRouter)
	if !ok {
		// Why not meaning it is only 1?
	}
	if len(router.Consumers() < 2 {
		return fmt.Errorf("routing connector requires at least two downstream pipelines")
		// or maybe we decide it's ok to have only one. point is it can be checked either way
	}
}
```